crates/noirc_frontend/src/lexer/token.rs

```diff
@@ -419,64 +447,107 @@ impl Attribute {
                     });
                 }
 
-                Attribute::Deprecated(name.trim_matches('"').to_string().into())
-            }
-            ["test"] => Attribute::Test(TestScope::None),
-            ["test", name] => {
-                validate(name)?;
-                let malformed_scope =
-                    LexerErrorKind::MalformedFuncAttribute { span, found: word.to_owned() };
-                match TestScope::lookup_str(name) {
-                    Some(scope) => Attribute::Test(scope),
-                    None => return Err(malformed_scope),
-                }
+                Attribute::Secondary(SecondaryAttribute::Deprecated(
+                    name.trim_matches('"').to_string().into(),
+                ))
             }
             tokens => {
                 tokens.iter().try_for_each(|token| validate(token))?;
-                Attribute::Custom(word.to_owned())
+                Attribute::Secondary(SecondaryAttribute::Custom(word.to_owned()))
             }
         };
 
         Ok(Token::Attribute(attribute))
     }
+}
 
+/// Primary Attributes are those which a function can only have one off
+/// They change the FunctionKind and thus have direct impact on the IR output
+#[derive(PartialEq, Eq, Hash, Debug, Clone, PartialOrd, Ord)]
+pub enum PrimaryAttribute {
+    Foreign(String),
+    Builtin(String),
+    Oracle(String),
+    Test(TestScope),
+}
+
+impl PrimaryAttribute {
     pub fn builtin(self) -> Option<String> {
         match self {
-            Attribute::Builtin(name) => Some(name),
+            PrimaryAttribute::Builtin(name) => Some(name),
             _ => None,
         }
     }
 
     pub fn foreign(self) -> Option<String> {
         match self {
-            Attribute::Foreign(name) => Some(name),
+            PrimaryAttribute::Foreign(name) => Some(name),
             _ => None,
         }
     }
 
     pub fn is_foreign(&self) -> bool {
-        matches!(self, Attribute::Foreign(_))
+        matches!(self, PrimaryAttribute::Foreign(_))
     }
 
     pub fn is_low_level(&self) -> bool {
-        matches!(self, Attribute::Foreign(_) | Attribute::Builtin(_))
+        matches!(self, PrimaryAttribute::Foreign(_) | PrimaryAttribute::Builtin(_))
+    }
+}
+
+impl fmt::Display for PrimaryAttribute {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PrimaryAttribute::Test(scope) => write!(f, "#[test{}]", scope),
+            PrimaryAttribute::Foreign(ref k) => write!(f, "#[foreign({k})]"),
+            PrimaryAttribute::Builtin(ref k) => write!(f, "#[builtin({k})]"),
+            PrimaryAttribute::Oracle(ref k) => write!(f, "#[oracle({k})]"),
+        }
+    }
+}
+
+/// Secondary attributes are those which a function can have many off
+/// They do not change the FunctionKind and thus do not have direct impact on the IR output
```

Comment:
nit: probably should say they are not able to change the FunctionKind, tests currently don't change the FunctionKind but perhaps with this new abstraction, they should